### PR TITLE
RUMM-156 Introduce the GitDiffsConditionalDependencyPlugin

### PR DESCRIPTION
**Pull request by mariusc83**
### What does this PR do?

This commit introduce the *GitDiffsConditionalDependencyPlugin* which provides the *GitDiffsTask* as an alternative to the default gradle tasks. 
Using the *GitDiffsTask* you will be able to add dependencies to it conditioned by git differences with the target branch as :
`gitDiffTask("unitTestAll") {
    dependsOnDiff(
        "dd-sdk-android/.*",
        ":dd-sdk-android:testDebugUnitTest",
        ":dd-sdk-android:testReleaseUnitTest",
        ":sample:java:assembleDebug",
        ":sample:kotlin:assembleDebug"
    )
}`

### Motivation

We needed to lower the build time on our CI servers and to avoid executing tasks that are not relevant for the current PR.

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: buildSrc/src/main/kotlin/com/datadog/gradle/plugin/gitdiffs/GitDiffsExtension.kt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package com.datadog.gradle.plugin.gitdiffs
+
+import org.gradle.api.Task
+
+class GitDiffsExtension(val deps: MutableMap<Regex, MutableList<Dependency>> = mutableMapOf()) {
```
Comment (xgouchet) [inline buildSrc/src/main/kotlin/com/datadog/gradle/plugin/gitdiffs/GitDiffsExtension.kt]: I may have missed something but it seems you never use this class anywhere :/

File: buildSrc/src/main/kotlin/com/datadog/gradle/plugin/gitdiffs/GitDiffsProvider.kt [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+    open val diffs: List<String> by lazy {
+        val currentBranch = currentBranch()
+        val targetBranch = targetBranch()
+        println("Current branch $currentBranch")
+        project.execShell("git", "diff", "--name-only", currentBranch, targetBranch)
+    }
+
+    private fun targetBranch(): String {
+        val targetBranch = System.getProperty("ghprbTargetBranch")
```
Comment (xgouchet) [inline buildSrc/src/main/kotlin/com/datadog/gradle/plugin/gitdiffs/GitDiffsProvider.kt]: Is this property something specific to the plugin you inspired yourself from ?  Maybe there's an evn variable that's accessible on gitlab CI's containers that would make more sense using ?

File: buildSrc/src/main/kotlin/com/datadog/gradle/plugin/gitdiffs/GitDiffsExtension.kt (hunk 1 above)
Comment (mariusc83) [reply buildSrc/src/main/kotlin/com/datadog/gradle/plugin/gitdiffs/GitDiffsExtension.kt]: hmmmm I used to use it in the plugin, I will check

File: buildSrc/src/main/kotlin/com/datadog/gradle/plugin/gitdiffs/GitDiffsProvider.kt (hunk 2 above)
Comment (mariusc83) [reply buildSrc/src/main/kotlin/com/datadog/gradle/plugin/gitdiffs/GitDiffsProvider.kt]: That's what I wanted to ask you today. If you know a way we can get this from Gitlab.

File: buildSrc/src/main/kotlin/com/datadog/gradle/plugin/gitdiffs/GitDiffConditionalDependencyTask.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        val list = dependencies[regEx] ?: mutableListOf<String>()
+            .apply { dependencies[regEx] = this }
+        tasks.forEach {
+            list.add(it)
+        }
+    }
+}
+
+fun Project.gitDiffTask(name: String, configuration: Action<GitDiffConditionalDependencyTask>) {
+    tasks.register("unitTestAll", GitDiffConditionalDependencyTask::class, configuration)
```
Comment (mariusc83) [inline buildSrc/src/main/kotlin/com/datadog/gradle/plugin/gitdiffs/GitDiffConditionalDependencyTask.kt]: you forgot to use the _name_ here

**Pull request merged**